### PR TITLE
[CA-1790] Get New Google Project from RBS for v2 Workspaces

### PR DESCRIPTION
**Pull request by ehigham**
RR: https://broadworkbench.atlassian.net/browse/CA-1790
Allow the migration actor to issue its own idempotence token to RBS when
claiming a google project via the WorkspaceService.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 1]
```diff
@@ -1967,27 +1967,31 @@ class WorkspaceService(protected val userInfo: UserInfo,
...
+                         rbsHandoutRequestId: String,
+                         span: Span = null) : Future[(GoogleProjectId, GoogleProjectNumber)] =
     for {
       googleProjectId <- traceWithParent("getGoogleProjectFromBuffer", span) { _ =>
-        resourceBufferService.getGoogleProjectFromBuffer(projectPoolType, workspaceId)
+        resourceBufferService.getGoogleProjectFromBuffer(
+          if (billingProject.servicePerimeter.isDefined)
+            ProjectPoolType.ExfiltrationControlled else
+            ProjectPoolType.Regular,
+          workspaceId
```
Comment (s-rubenstein) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Should this be `rbsHandoutRequestId`?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -1967,27 +1967,31 @@ class WorkspaceService(protected val userInfo: UserInfo,
...
-
+                         rbsHandoutRequestId: String,
+                         span: Span = null) : Future[(GoogleProjectId, GoogleProjectNumber)] =
     for {
       googleProjectId <- traceWithParent("getGoogleProjectFromBuffer", span) { _ =>
-        resourceBufferService.getGoogleProjectFromBuffer(projectPoolType, workspaceId)
+        resourceBufferService.getGoogleProjectFromBuffer(
+          if (billingProject.servicePerimeter.isDefined)
+            ProjectPoolType.ExfiltrationControlled else
+            ProjectPoolType.Regular,
```
Comment (andy7i) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1985]: nit: in-lining this makes it a bit harder to read

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -1967,27 +1967,31 @@ class WorkspaceService(protected val userInfo: UserInfo,
...
+          workspaceId
+        )
       }
 
       _ = logger.info(s"Moving google project ${googleProjectId} to folder.")
       _ <- traceWithParent("maybeMoveGoogleProjectToFolder", span) { _ =>
-        maybeMoveGoogleProjectToFolder(billingProject.servicePerimeter, googleProjectId)
+        billingProject.servicePerimeter.traverse_ {
+          userServiceConstructor(userInfo).moveGoogleProjectToServicePerimeterFolder(_, googleProjectId)
+        }
```
Comment (andy7i) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1994]: nit: I'm not familiar with `traverse_` so it's a bit harder for me to validate this bit of code just from looking at it. 
(without knowing what `traverse_` does, I can't reason whether `moveGoogleProjectToServicePerimeterFolder` will get called or not when `billingProject.servicePerimeter` is empty)

**Review by andy7i — COMMENTED** [review]
drive-by review

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 1 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: yes. It should. Thank you :)

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 2 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1985]: I actually find it easier to read as it reduces the number variables in scope that I have to keep track of while reading the code.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 3 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1994]: `traverse` is quite a common abstraction in FP and has similar behaviour to `map` - here's a short video that introduces the `Traverse` type class in scala that gives a bunch of examples: https://www.youtube.com/watch?v=CSErp1mZodM 
Hint on naming conventions: if you see a function name that ends in `_`, it's likely to ignore the output and return `F[Unit]`, for some applicative `F`.

**Review by gpolumbo-broad — APPROVED** [review]

**Review by s-rubenstein — APPROVED** [review]
LGTM

**Pull request merged**